style_guides/angular_style_guide.md

```diff
@@ -888,6 +888,144 @@ Kibana is written in Angular, and these are the conventions we follow. Most of t
   <div>min={{vm.min}}<input ng-model="vm.min"/></div>
   ```
 
+### Use callbacks to communicate side effects up the hierarchy
+
+  - Treat attribute bindings as read-only and communicate with with the parent
+    directive via callbacks. Essentially treat directive attribute bindings
+    like function arguments, which should also not be mutated.
+
+    *Why?*: Spreading mutable state throughout the application makes it very
+    difficult to understand and predict. By receiving data from the parent
+    through the bindings and communicating desired changes back via separate
+    callback attributes the data flow becomes explicit. The attribute bindings
+    then become the public interface of the component, describing all the
+    inputs dictating the rendering and all possible effects on the application
+    state.
+
+  ```javascript
+  /* avoid */
+  module.directive('myBadTableHeader', function MyBadTableHeader() {
+    return {
+      bindToController: true,
+      controllerAs: 'tableHeader',
+      controller: TableHeaderController,
+      replace: true,
+      scope: {
+        caption: '=',
+        sortColumn: '=',
+      },
+      template: `<th
+        ng-class="{ 'tableHeader--sortColumn': tableHeader.isSortColumn() }"
+        ng-bind="tableHeader.caption"
+        ng-click="tableHeader.sort()"
+      ></th>`,
+    };
+  });
+
+  function TableHeaderController() {
+    this.isSortColumn = () => (
+      this.sortColumn === this.caption
+    );
+
+    this.sort = () => (
+      this.sortColumn = this.caption  // AVOID: mutating input attributes
+    );
+
+    /* OTHER STUFF */
+  }
+  ```
+
+  ```javascript
+  /* recommended */
+  module.directive('myGoodTableHeader', function MyGoodTableHeader() {
+    return {
+      bindToController: true,
+      controllerAs: 'tableHeader',
+      controller: TableHeaderController,
+      replace: true,
+      scope: {
+        caption: '=',        // INPUT
+        isSortColumn: '=',   // INPUT
+        setSortColumn: '=',  // EFFECT
```

Comment:
Yes, it would make sense to settle on one naming convention. Two options come to mind: the `on` prefix or using a strong verb (as is usually recommended for all function names). After 5 minutes of thought the differences seem to be:

* `on` focuses on what the origin of the effect in the consuming directive is
* the verb focuses on what the intended effect is

For the sake of decoupling the `on` sounds more sensible. What the actual effect is should not concern the component itself. That also seems to be the convention used by React.